File: 39.组合总和.java
import java.util.Arrays;
import java.util.Deque;

/*
 * @lc app=leetcode.cn id=39 lang=java
 *
 * [39] 组合总和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        int len = candidates.length;
        Arrays.sort(candidates);
        dfs(candidates, len, target, 0, new ArrayDeque<>(), res);

        return res;
    }

    private void dfs(int[] candidates, int len, int residue, int begin, Deque<Integer> path, List<List<Integer>> res){
        if (residue == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = begin; i < len; i++) {
            if (residue - candidates[i] < 0) break;
            path.addLast(candidates[i]);
            dfs(candidates, len, residue - candidates[i], i, path, res);
            path.removeLast();
        }

    }

}
// @lc code=end

File: 39.组合总和.py
# @before-stub-for-debug-begin
from python3problem39 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode.cn id=39 lang=python3
#
# [39] 组合总和
#

# @lc code=start
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        size = len(candidates)
        ans = []
        def solve(rec, index, new_target):
            if new_target == 0:
                # print("rec is", rec)
                ans.append(list(rec))
                # print("ans is", ans)
                return
            for i in range(index, size):
                if candidates[i] <= new_target:
                    rec.append(candidates[i])
                    solve(rec, i, new_target-candidates[i])
                    rec.pop()
            
        solve([], 0, target)
        return ans

    
# @lc code=end

